File: LeetCode/Medium/695 - Max Area of Island/695 - Max Area of Island.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
 var maxAreaOfIsland = function(grid) {
    const n = grid.length, m = grid[0].length;
    const dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    
    const valid = (i, j) => i >= 0 && i < n && j >= 0 && j < m;

    const DFS = (i, j) => {
        if (!valid(i, j) || !grid[i][j]) return 0;
        grid[i][j] = 0;

        let ret = 1;
        for (let dir of dirs) ret += DFS(i + dir[0], j + dir[1]);

        return ret;
    }

    let ret = 0;
    for (let i = 0; i < n; ++i) {
        for (let j = 0; j < m; ++j) {
            if (grid[i][j]) 
                ret = Math.max(ret, DFS(i, j));
        }
    }
    
    return ret;
};

File: LeetCode/Medium/695 - Max Area of Island/695 - Max Area of Island.cpp
class Solution {
private:
    vector<int> dx = {0,  0, 1, -1};
    vector<int> dy = {1, -1, 0,  0};
    int n, m;
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        n = grid.size(), m = grid[0].size();
        int mx = 0;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (grid[i][j]) {
                    mx = max(mx, DFS(grid, i, j));
                }
            }
        }
        return mx;
    }
    
    int DFS(vector<vector<int>> &grid, int i, int j) {
        if (!valid(i, j) || !grid[i][j]) return 0;
        grid[i][j] = 0;
        
        int ret = 1;
        for (int k = 0; k < 4; ++k) {
            ret += DFS(grid, i + dx[k], j + dy[k]);
        }
        return ret;
    }
    
    bool valid(int i, int j) {
        return i >= 0 && i < n && j >= 0 && j < m;
    }
};